src/lib.rs

```diff
@@ -2528,34 +2528,63 @@ pub trait Itertools: Iterator {
-    /// - otherwise if `f` is non-trivial like `format!`, you should use `tree_reduce` since it
-    /// reduces the number of operations from `O(n)` to `O(ln(n))`
+    /// For an iterator producing `n` elements, both [`Iterator::reduce`] and `tree_reduce` will
+    /// call `f` `n - 1` times. However, `tree_reduce` will call `f` on earlier intermediate
+    /// results, which is beneficial for `f` that allocate and produce longer results for longer
+    /// arguments. For example if `f` combines arguments using `format!`, then `tree_reduce` will
+    /// operate on average on shorter arguments resulting in less bytes being allocated overall.
     ///
-    /// Here "non-trivial" means:
-    ///
-    /// - any allocating operation
-    /// - any function that is a composition of many operations
+    /// If 'f' does not benefit from such a reordering, like `u32::wrapping_add`, prefer the
+    /// normal [`Iterator::reduce`] instead since it will most likely result in the generation of
+    /// simpler code because the compiler is able to optimize it.
     ///
     /// ```
     /// use itertools::Itertools;
     ///
+    /// let f = |a: String, b: String| {
+    ///     format!("f({a}, {b})")
+    /// };
+    ///
     /// // The same tree as above
-    /// let num_strings = (1..8).map(|x| x.to_string());
-    /// assert_eq!(num_strings.tree_reduce(|x, y| format!("f({}, {})", x, y)),
-    ///     Some(String::from("f(f(f(1, 2), f(3, 4)), f(f(5, 6), 7))")));
+    /// assert_eq!((1..8).map(|x| x.to_string()).tree_reduce(f),
+    ///            Some(String::from("f(f(f(1, 2), f(3, 4)), f(f(5, 6), 7))")));
     ///
     /// // Like fold1, an empty iterator produces None
     /// assert_eq!((0..0).tree_reduce(|x, y| x * y), None);
     ///
     /// // tree_reduce matches fold1 for associative operations...
     /// assert_eq!((0..10).tree_reduce(|x, y| x + y),
     ///     (0..10).fold1(|x, y| x + y));
+    ///
     /// // ...but not for non-associative ones
     /// assert_ne!((0..10).tree_reduce(|x, y| x - y),
     ///     (0..10).fold1(|x, y| x - y));
+    ///
+    ///
+    /// let mut total_capacity_reduce = 0;
+    /// let reduce_res = (1..100).map(|x| x.to_string())
+    ///     .reduce(|a, b| {
+    ///         let r = f(a, b);
+    ///         total_capacity_reduce += r.capacity();
+    ///         r
+    ///     })
+    ///     .unwrap();
+    ///
+    /// let mut total_capacity_tree_reduce = 0;
+    /// let tree_reduce_res = (1..100).map(|x| x.to_string())
+    ///     .tree_reduce(|a, b| {
+    ///         let r = f(a, b);
+    ///         total_capacity_tree_reduce += r.capacity();
+    ///         r
+    ///     })
+    ///     .unwrap();
+    ///
+    /// dbg!(total_capacity_reduce, total_capacity_tree_reduce,
+    ///      reduce_res.len(), tree_reduce_res.len());
+    /// // total_capacity_reduce = 65630
+    /// // total_capacity_tree_reduce = 7284
+    /// // reduce_res.len() = 679
+    /// // tree_reduce_res.len() = 679
```

Comment:
We usually don't use `dbg!` but assertions. These comments could silently be outdated later. But I guess there is no guarantees on allocated capacities for such example? I'm not sure if it's better to have assertions that might fail in the future because of changed allocations in future Rust.

Reply:
Perhaps we can create a mock, hidden version of `String` , here? I don't want us to have fragile assertions.

Reply:
Its probably useful to have some example that shows the possible runtime improvements of `tree_reduce` over `reduce`, because the text itself is quite abstract. But maybe we can use another example that doesn't involve `String` and is deterministic?

Reply:
I changed the example to using `.len()` instead of `.capacity()` which is deterministic and changed the `dbg!` to `assert!`. What do you think about that?

Reply:
Yeah, I thought `capacity` was somehow needed over `len` but it seems more appropriate.